File: 0202-happy-number-快乐数/0202.cpp
class Solution {
    int calc(int n) {
        int rtn = 0;
        while (n > 0) {
            int m = n % 10;
            rtn += m * m;
            n /= 10;
        }
        return rtn;
    }
public:
    bool isHappy(int n) {
        int fast = n, slow = n;
        while (fast != 1 || slow != 1) {
            // 慢指针计算一次
            slow = calc(slow);
            // 快指针计算两次
            fast = calc(fast);
            fast = calc(fast);
            // std::cout << fast << "," << slow << std::endl;
            if (fast == slow) {
                if (fast == 1) {
                    return true;
                }
                return false;
            }
        }
        return true;
    }
};

File: 0202-happy-number-快乐数/0202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        t = n
        resultDict = {}
        while True:
            if t == 1: return True
            if t in resultDict:
                return False
            else: resultDict[t] = False
            r = 0
            while t > 0:
                m = (t % 10) * (t % 10)
                r += m
                t = t // 10
            t = r

        return False
